### PR TITLE
Pregel additional test & TSan error fix

### PR DESCRIPTION
**Pull request by graetzer**
### Scope & Purpose

Add a test for the 'hits' pregel algorithm, fix a lock-inversion warning when running with the thread sanitizer

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] The behaviour change can only be verified via automatic tests

### Testing & Verification

This change is already covered the existing pregel tests

- [x] Added new **integration tests** (i.e. in shell_server / shell_server_aql)

Additionally:

- [x] I ensured this code runs with ASan / TSan or other static verification tools

http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/6601/

### CODE REVIEW

File: arangod/Pregel/PregelFeature.cpp [hunk 1]
```diff
@@ -316,12 +316,16 @@ void PregelFeature::cleanupWorker(uint64_t executionNumber) {
 
 void PregelFeature::cleanupAll() {
   MUTEX_LOCKER(guard, _mutex);
-  _conductors.clear();
-  for (auto it : _workers) {
+  decltype(_conductors) cs = std::move(_conductors);
+  decltype(_workers) ws = std::move(_workers);
+  guard.unlock();
+    
+  cs.clear();
```
Comment (jsteemann) [inline arangod/Pregel/PregelFeature.cpp]: mmh. `cs` is a copy/move of `_conductors`, but is instantly cleared afterwards. Is that intentional?

File: arangod/Pregel/PregelFeature.cpp [hunk 2]
```diff
@@ -316,12 +316,16 @@ void PregelFeature::cleanupWorker(uint64_t executionNumber) {
...
+  for (auto it : ws) {
     it.second.second->cancelGlobalStep(VPackSlice());
   }
   std::this_thread::sleep_for(std::chrono::microseconds(1000 * 100));  // 100ms to send out cancel calls
-  _workers.clear();
+  cs.clear();
```
Comment (jsteemann) [inline arangod/Pregel/PregelFeature.cpp]: `cs` will already be empty here.

File: arangod/Pregel/PregelFeature.cpp (hunk 1 above)
Comment (danielhlarkin) [reply arangod/Pregel/PregelFeature.cpp]: Looks like it. The old code simply cleared it under the lock. This way the deallocation can be done outside the lock, while the move itself should likely be quicker.

File: tests/js/server/shell/shell-pregel.js [hunk 3]
```diff
@@ -383,6 +383,28 @@ function randomTestSuite() {
...
+      };
+      var pid = pregel.start("hits", graphName, opts);
+      var i = 10000;
+      do {
+        internal.wait(0.2);
+        var stats = pregel.status(pid);
+        if (stats.state !== "running") {
+          assertEqual(stats.vertexCount, n, stats);
+          assertEqual(stats.edgeCount, m * 2, stats);
+          db
```
Comment (danielhlarkin) [inline tests/js/server/shell/shell-pregel.js]: Is this `db` a typo? Missing code?

Comment (jsteemann) [inline tests/js/server/shell/shell-pregel.js]:
```suggestion
```

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5000/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5054/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5060/

**Pull request merged**